### PR TITLE
Fixing precision error for stretched SliverAppBars

### LINKED ISSUE #53823: Anti-aliasing apparent in Stretched FlexibleSpace.title
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update `main.dart` file as follows:
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    MaterialApp(
      home: Scaffold(
        body: CustomScrollView(
          controller: ScrollController(),
          physics: const BouncingScrollPhysics(),
          slivers: <Widget>[
            SliverAppBar(
              elevation: 0.0,
              expandedHeight: 200.0,
              pinned: false,
              stretch: true,
              leading: const CloseButton(color: Colors.white),
              flexibleSpace: FlexibleSpaceBar(
                background: Image.network(
                  'https://upload.wikimedia.org/wikipedia/commons/thumb/b/b2/Panor%C3%A1mica_Elche.JPG/1920px-Panor%C3%A1mica_Elche.JPG',
                  fit: BoxFit.cover,
                ),
                collapseMode: CollapseMode.none,
                title: Container(
                  height: 8.0,
                  decoration: const BoxDecoration(
                    borderRadius: BorderRadius.vertical(
                      top: Radius.circular(20.0),
                    ),
                    color: Colors.white,
                  ),
                ),
                centerTitle: true,
                titlePadding: const EdgeInsets.all(0),
              ),
            ),
            SliverList(
              delegate: SliverChildListDelegate(
                <Widget>[
                  ColoredBox(
                    color: Colors.white,
                    child: Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 16.0),
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.start,
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: const <Widget>[
                          SizedBox(height: 16.0),
                          Placeholder(),
                          SizedBox(height: 16.0),
                          Placeholder(),
                          SizedBox(height: 16.0),
                          Placeholder(),
                          SizedBox(height: 16.0),
                        ],
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    ),
  );
}
```
3. Scroll in any direction.

- If you scroll from top to bottom, while overscrolling thanks to `stretch: true`, you can notice a weird line appears above the `Placeholder`s (between `SliverAppBar` and `SliverList`). 
- If you scroll from bottom to top, a _smaller_ line appears. If you stop scrolling, the line will remain rendered.

**Expected results:** No gap between between `SliverAppBar` and `SliverList`.

**Actual results:** A weird line appears between `SliverAppBar` and `SliverList`.

<details>
  <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, v1.16.4-pre.69, on Microsoft Windows [VersiÃ³n 10.0.18363.752], locale es-ES)
    • Flutter version 1.16.4-pre.69 at C:\flutter
    • Framework revision 19e7db585d (18 hours ago), 2020-04-01 11:16:01 -0700
    • Engine revision c9506cb8e9
    • Dart version 2.8.0 (build 2.8.0-dev.18.0 eea9717938)


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at C:\Users\crist\AppData\Local\Android\sdk
    • Platform android-R, build-tools 29.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[✓] Android Studio (version 3.6)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 45.0.2-dev.5
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)

[✓] Connected device (3 available)
    • Pixel 3 XL • 192.168.1.33:5555 • android-arm64  • Android 10 (API 29)
    • Chrome     • chrome            • web-javascript • Google Chrome 80.0.3987.149
    • Web Server • web-server        • web-javascript • Flutter Tools

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by Piinks**
## Description

This fixes a precision error that is apparent when aligning elements in the `Stack` of a `FlexibleSpace` along the bottom of the stretched `SliverAppBar`.

## Related Issues

Fixes #53823 

## Tests

Stretch fills overscroll and handles precision error.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/sliver_persistent_header.dart [hunk 1]
```diff
@@ -178,7 +178,7 @@ abstract class RenderSliverPersistentHeader extends RenderSliver with RenderObje
     }());
     double stretchOffset = 0.0;
     if (stretchConfiguration != null && childMainAxisPosition(child) == 0.0)
-      stretchOffset += constraints.overlap.abs();
+      stretchOffset += constraints.overlap.abs().floorToDouble();
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/sliver_persistent_header.dart:181]: usually, we fix this error by using https://api.flutter.dev/flutter/foundation/precisionErrorTolerance-constant.html
in the `if` condition check

Comment (Piinks) [reply packages/flutter/lib/src/rendering/sliver_persistent_header.dart:181]: So I've tried this with the precisionErrorTolerance constant, and it looks like the issue still persists. I think we'll need a value higher than the current constant for this bug.

Comment (Piinks) [reply packages/flutter/lib/src/rendering/sliver_persistent_header.dart:181]: Here is a clip: 

ezgif com-video-to-gif (4)

Comment (chunhtai) [reply packages/flutter/lib/src/rendering/sliver_persistent_header.dart:181]: I apply this fix locally, but i still see the same issue. Am i doing something wrong?
sliverheader

Comment (Piinks) [reply packages/flutter/lib/src/rendering/sliver_persistent_header.dart:181]: Oh wow. That is bizarre. Let me try it again locally.

Comment (Piinks) [reply packages/flutter/lib/src/rendering/sliver_persistent_header.dart:181]: Stranger still. It looks like the fix works on an iPhone 11 simulator, but not a Pixel 2 emulator. I'll continue to investigate.

**Comment by Piinks** [comment]
@HansMuller suggested we are seeing an anti-aliasing issue here.

**Comment by HansMuller** [comment]
This does look like the kind of antialiasing issue that's introduced by trying to stack containers in one color on a background with a strongly contrasting color.  The problem occurs because of the mapping from logical pixels to device pixels. There's an open issue for a documentation fix: #14288

More info about the problem can be found here:
#14288 (comment)
#17084 (comment)
#15035 (comment)

**Comment by goderbauer** [comment]
@liyuqian Do you have any thoughts on this (potential) AA issue?

**Comment by liyuqian** [comment]
@goderbauer : ah yes, it's a known issue and some of the workarounds are discussed in https://github.com/flutter/flutter/issues/25009

**Comment by HansMuller** [comment]
@goderbauer @liyuqian - Is this new issue another example of the antialiasing problem? https://github.com/flutter/flutter/issues/54370#issuecomment-611714132

**Comment by liyuqian** [comment]
I believe so. If such issue would occur by using a built-in widget like `ToggleButton` (https://github.com/flutter/flutter/issues/54370), then we probably need to improve our widget implementation to avoid such artifacts.

**Comment by Piinks** [comment]
Since the issue appears to not be fixed by this change, I am going to close this. I'll update the bug.

**Pull request closed without merging**